### PR TITLE
[WIP] First pass at updating API version

### PR DESCRIPTION
**Pull request by andrewazores**
This successfully builds and bundles, but still needs to undergo testing.

### CODE REVIEW

**Comment by andrewazores** [comment]
`make test` passses.

**Comment by andrewazores** [comment]
`make image && podman push $IMAGE_TAG && podman image prune && make deploy sample_app2` shows good looking logs and everything passes a manual smoketest.

**Comment by andrewazores** [comment]
I created a bundle on my personal Quay namespace as operator version 0.4.0, and I cannot create a ContainerJFR instance through the OpenShift console. Need to investigate.

File: bundle/container-jfr-operator.v0.3.1.clusterserviceversion.yaml [hunk 1]
```diff
@@ -205,11 +205,13 @@ spec:
                 serviceAccountName: container-jfr-operator
                 containers:
                   - name: container-jfr-operator
-                    image: 'quay.io/rh-jmc-team/container-jfr-operator:0.3.0'
+                    image: 'quay.io/andrewazores/container-jfr-operator:0.3.1'
```
Comment (ebaron) [inline bundle/container-jfr-operator.v0.3.1.clusterserviceversion.yaml]: This should go back to rh-jmc-team when you're done with local testing.

File: bundle/openshift/operator-source.yaml [hunk 2]
```diff
@@ -6,6 +6,6 @@ metadata:
 spec:
     type: appregistry
     endpoint: https://quay.io/cnr
-    registryNamespace: rh-jmc-team
+    registryNamespace: andrewazores
```
Comment (ebaron) [inline bundle/openshift/operator-source.yaml]: Here too

File: go.mod [hunk 3]
```diff
@@ -34,3 +34,5 @@ replace (
 )
 
 replace github.com/operator-framework/operator-sdk => github.com/operator-framework/operator-sdk v0.11.0
+
+go 1.13
```
[the remark this answers is not in the document]
Comment (andrewazores) [reply go.mod]: Yea, I don't know either. I'll pull it back out of this patch, but we should probably do something about it separately.

File: bundle/container-jfr-operator.v0.3.1.clusterserviceversion.yaml (hunk 1 above)
Comment (andrewazores) [reply bundle/container-jfr-operator.v0.3.1.clusterserviceversion.yaml]: Yep.

**Comment by andrewazores** [comment]
I haven't gotten back around to testing this when uploaded to Quay as an application and installed using the OperatorHub UI and a custom source. I'll wait until Jie is done with his FOSDEM demo since the old build he's using for it is currently in my namespace. Deploying directly with oc and the make targets looks like it works perfectly fine though, so if there are any issues remaining they're "simple" ones in the bundle/* YAMLs.

**Comment by andrewazores** [comment]
@ebaron have you made any inroads here, or at least learned anything helpful, from using your tool to download and inspect the bundles from quay? I can come back around to looking at this issue again and see if we can update the CRDs properly, but I think you might have a better grip on this than I do right now.

**Comment by ebaron** [comment]
I haven't figured out the cause of it, but I have confirmed that the bundle manifest on Quay has the old CRDs present, but what operator-courier builds and sends to Quay has the new CRDs.

This one was built with the script found here: https://github.com/rh-jmc-team/container-jfr-operator/wiki/Inspecting-Application-Bundles
[bundle-local.yaml.txt](https://github.com/rh-jmc-team/container-jfr-operator/files/4168170/bundle-local.yaml.txt)

This version was pulled from Quay using helm using the wiki instructions.

I thought maybe it was just a bad build somehow, so I tried pushing again, but got the following:
```
ERROR:operatorcourier.push:{"error":{"code":"forbidden","details":{"package":"rh-jmc-team/container-jfr-operator-bundle","scopes":["push"]},"message":"Unauthorized access for: rh-jmc-team/container-jfr-operator-bundle"}}

Unauthorized access for: rh-jmc-team/container-jfr-operator-bundle 
```
I recall you saying that you can't overwrite versions. Is this also the error you got?

I noticed that you can use Helm to specify a `force` option when pushing, that you can't do with operator-courier. It might be worthwhile to build another 0.3.0 bundle locally, and trying to force push that to Quay and seeing if the problem persists. What do you think?
[bundle-quay.yaml.txt](https://github.com/rh-jmc-team/container-jfr-operator/files/4168171/bundle-quay.yaml.txt)

**Comment by andrewazores** [comment]
Interesting, that force push sounds like it could be handy.

That error message when pushing looks unfamiliar to me though. I see you have Write permissions on the Quay bundle "application", so I'm not sure what the problem is. You're using an account named "ebaron", right? I'll try elevating your permissions too.

**Comment by andrewazores** [comment]
helm only seems to push helm charts, not YAML CSVs like we have - unless I'm missing something. So I don't think its force-push option will be useful after all :-/

**Comment by andrewazores** [comment]
@ebaron I think operator-courier can actually do something similar for locally built bundles to what your python script does:

https://github.com/operator-framework/operator-courier/pull/91

**Comment by andrewazores** [comment]
Oh, re. the unauthorized push - did you have your `AUTH_TOKEN` env var set?

**Comment by ebaron** [comment]
I did not, I passed the token via command line argument like:
`operator-courier push bundle/ rh-jmc-team container-jfr-operator-bundle 0.3.0 my_token`

**Comment by ebaron** [comment]
> @ebaron I think operator-courier can actually do something similar for locally built bundles to what your python script does:
> 
> [operator-framework/operator-courier#91](https://github.com/operator-framework/operator-courier/pull/91)

Oh interesting. I didn't see that in the `--help` for verify, and I must have missed it when looking at the code.

**Comment by andrewazores** [comment]
Oh... nevermind.

https://github.com/operator-framework/operator-courier/commit/a2a6652bdcdcb06aac8e02c258bd87012c48e18b

It got reverted shortly after it was added.

**Comment by andrewazores** [comment]
I just built a new 0.3.1 bundle with the nested structure and pushed it to my andrewazores quay namespace, and when I use the helm registry plugin to pull it back down, the CRDs are correct. So this might solve our problem. I'm going to try installing the operator from that namespace as a custom source in my crc instance and see what I get.

**Comment by andrewazores** [comment]
Looks like no dice:

```
{"level":"info","ts":1581098567.5990236,"logger":"controller-runtime.controller","msg":"Starting EventSource","controller":"containerjfr-controller","source":"kind source: /, Kind="}
{"level":"error","ts":1581098571.0434058,"logger":"controller-runtime.source","msg":"if kind is a CRD, it should be installed before calling Start","kind":"ContainerJFR.rhjmc.redhat.com","error":"no matches for kind \"ContainerJFR\" in version \"rhjmc.redhat.com/v1alpha1\"","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/andrew/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/source.(*Kind).Start\n\t/home/andrew/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.2/pkg/source/source.go:88\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Watch\n\t/home/andrew/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.2/pkg/internal/controller/controller.go:122\ngithub.com/rh-jmc-team/container-jfr-operator/pkg/controller/containerjfr.add\n\tcontainer-jfr-operator/pkg/controller/containerjfr/containerjfr_controller.go:50\ngithub.com/rh-jmc-team/container-jfr-operator/pkg/controller/containerjfr.Add\n\tcontainer-jfr-operator/pkg/controller/containerjfr/containerjfr_controller.go:33\ngithub.com/rh-jmc-team/container-jfr-operator/pkg/controller.AddToManager\n\tcontainer-jfr-operator/pkg/controller/controller.go:13\nmain.main\n\tcontainer-jfr-operator/cmd/manager/main.go:118\nruntime.main\n\t/usr/lib/golang/src/runtime/proc.go:203"}
{"level":"error","ts":1581098571.0441027,"logger":"cmd","msg":"","error":"no matches for kind \"ContainerJFR\" in version \"rhjmc.redhat.com/v1alpha1\"","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/andrew/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nmain.main\n\tcontainer-jfr-operator/cmd/manager/main.go:119\nruntime.main\n\t/usr/lib/golang/src/runtime/proc.go:203"}
```

**Comment by andrewazores** [comment]
Not sure where it's still getting `v1alpha` from.

**Comment by ebaron** [comment]
So I managed to overwrite the 0.3.0 version on Quay using the force option manually with cURL.

```shell
$ BLOB=$(cat rh-jmc-team_container-jfr-operator-bundle_0.3.0.tar.gz | base64 -w 0)
$ curl -XPOST -H "Authorization: $TOKEN" -H "Content-Type: application/json" \
> https://quay.io/cnr/api/v1/packages/rh-jmc-team/container-jfr-operator-bundle?force=true \
> -d '{ "blob": "'"${BLOB}"'", "release": "0.3.0", "media_type": "helm" }'
{"status":"ok"}
```

Pulling the manifest from Quay shows the expected CRDs now.
```shell
$ helm registry pull quay.io/rh-jmc-team/container-jfr-operator-bundle
Pull package: rh-jmc-team/container-jfr-operator-bundle... 
./rh-jmc-team_container-jfr-operator-bundle_0.3.0
```

[bundle.yaml.txt](https://github.com/rh-jmc-team/container-jfr-operator/files/4172268/bundle.yaml.txt)

**Comment by andrewazores** [comment]
Ha, nice. I get why operator-courier doesn't want to expose that option for people to (ab)use for actually deployed things, but it sure would be nice to have for alpha testing. Much more convenient than having to delete and recreate the namespace every time, especially for testing version upgrades like this.

Did you overwrite it with a nested bundle or still a flat structured one?

**Comment by ebaron** [comment]
I should clarify the tarball was created by downloading the existing version from Quay with:
`$ helm registry pull --tarball quay.io/rh-jmc-team/container-jfr-operator-bundle` and replacing the bundle.yaml with a corrected one. It simply contains:
```
$ tar tzf rh-jmc-team_container-jfr-operator-bundle_0.3.0.tar.gz 
container-jfr-operator-bundle-fszkqf2v/
container-jfr-operator-bundle-fszkqf2v/bundle.yaml
```

**Comment by ebaron** [comment]
> Ha, nice. I get why operator-courier doesn't want to expose that option for people to (ab)use for actually deployed things, but it sure would be nice to have for alpha testing. Much more convenient than having to delete and recreate the namespace every time, especially for testing version upgrades like this.
> 
> Did you overwrite it with a nested bundle or still a flat structured one?

Still a flat-structured one.

**Pull request closed without merging**